### ISSUE TITLE
Allowing calling Meteor methods with an option to not retry them on hot-reload

### ISSUE THREAD

**Issue by mitar** [meteor/meteor#6108]
Currently, if there is any pending Meteor method being run and the server reloads (DDP connection is re-established) those Meteor methods are retried. It would be great if one could disable that for particular method calls.

**Comment by tmeasday** [comment]
Agreed, I think this would be a useful feature. I don't think we'll implement this in the current livedata implementation. @stubailo - would we take a PR that implemented it?

**Comment by stubailo** [comment]
Yeah this is something I've actually wanted for a while - the ability to completely disable retries, and throw an error if we don't know that the method succeeded, or have reason to believe it failed.

Would super accept a PR for this, and would love to make it the default in ValidatedMethod. (whether or not you are OK with retrying would depend on whether the method is idempotent)

**Comment by lassombra** [comment]
I'm interested in tackling this but have a question about how people would like to see this implemented.  As I see it, there are a couple of for sure details and a couple of options.

Defining a method as idempotent:

- Could be done in method definition.
- Could be an option passed to call

Handling idempotent with reconnect:
- Should throw an error.  
- Error should describe the situation
- Reconnect would ideally have a way to query the old subscription id and try to get the result of the method to handle dropped connection vs server reload.

Default setting:
- Should probably be methods are not idempotent (current behavior)
- Could be idempotent methods by default?

I am leaning towards defining in method definition, not at method call.  I am also leaning towards defaulting to non-idempotent for reliable backwards compatible behavior.

I've begun work on this and will submit a PR when done.

**Comment by stubailo** [comment]
I think you have the opposite meaning of idempotent. An idempotent method is one that is okay to call multiple times.

Canceling retries has to be an option passed to Meteor.apply because the client doesn't have any way of accessing definitions of server-only methods. Then we can integrate that option into a method calling wrapper like ValidatedMethod.

The default should be what happens in current meteor for backward compatibility reasons.

**Comment by tmeasday** [comment]
Oh I and I was going to add:

> Reconnect would ideally have a way to query the old subscription id and try to get the result of the method to handle dropped connection vs server reload.

Do you mean "old connection id"? This sounds like something that's out of scope; we can't guarantee the client will connect to the same web server and I don't believe there's any way to identify this connection as being the "same" as an old one (like what property would you reliably use?)

**Comment by lassombra** [comment]
Yeah, I'm tired and was thinking backwards and inside out.  
@stubailo, you make a good point about the client not knowing the definition.  So my ideal there would only be usable in a shared code situation, which obviously doesn't work.

@tmeasday, I did figure that would be out of scope, but figured I'd throw it out there.  The client does know what session ID it used before, and could possibly try to reuse that session ID.  Basically I was toying with the idea of fulfilling another issue that I saw mentioned which is that reconnects always end up with a completely new session.  I think I'll leave that out of this work (at least this PR that I'm making).

So basically:

Defining a method as idempotent:
- Could be done in method definition. (Not really possible without actually creating a gate of some kind on the method invocation, which is way out of scope.)
- Could be an option passed to call. (Definitely what I'll go with).

Handling _non_-idempotent with reconnect:
- Should throw an error.

Default setting:
- Should probably be methods are assumed to be idempotent (current behavior, what I'm going to go with)
- Could be non-idempotent methods by default? (breaking change, not recommended).

**Comment by mitar** [comment]
I would suggest that we do not complicate here and just implement this as an option to `Meteor.apply`, something like `Meteor.apply('foobar', {noRetry: true})` and this is it.

Then 3rd party libraries can provide wrappers around Meteor methods which do high-level semantics and map them to that option as needed. No need to push high-level semantics (which might differ between different use cases) down to Meteor core.

**Comment by lassombra** [comment]
@mitar, that's what I'm planning to do.  I am thinking about my own use cases and realize that it would be nice to be able to force it on the server, but that's impractical for now.

**Comment by stubailo** [comment]
I'd like to add this as an option to validated-method: https://github.com/meteor/validated-method

That way, you can specify the option when you _define_ the method, rather than when you call it. If anyone wants to PR that, I'd be happy to accept.

**Comment by rondip-geekyants** [comment]
Using noRetry:true, executes the stub only once, which is great, however an error is thrown saying `invocation-failed` : `Failed because noRetry option was passed`. How do I get rid of this? Please help. Thanks!
